File: fixtures/rules/AR001/safe/harnesskit_pattern_retry_on_contention.js
// Regression fixture (found via independent dogfood review against real HarnessKit code):
// scripts/review-independence/run-json-atomic.js's retryOnContention() was a confirmed false
// positive. A for(;;) retry loop whose bound is a compound condition
// (`!isContendedLockError(e) || Date.now() >= deadline`) exiting via `throw`, not a bare
// `if <comparison>: break/return`. Proves the widened has_step_bound heuristic
// (testContainsBoundComparison + throw-as-exit) now recognizes it as bounded.
function retryOnContention(fn, deadline) {
  for (;;) {
    try {
      return fn();
    } catch (e) {
      if (!isContendedLockError(e) || Date.now() >= deadline) throw e;
      sleepSync(50);
    }
  }
}

File: fixtures/rules/AR001/safe/harnesskit_pattern_retry_on_contention.py
# Regression fixture (found via independent dogfood review against real HarnessKit code):
# Python mirror of scripts/review-independence/run-json-atomic.js's retryOnContention() (JS),
# a confirmed false positive. A while True: retry loop whose bound is a compound condition
# (`not is_contended(e) or time.time() >= deadline`) exiting via `raise`, not a bare
# `if <comparison>: break/return`. Proves the widened _has_visible_step_bound heuristic
# (_test_contains_bound_comparison + raise-as-exit) now recognizes it as bounded.
def retry_on_contention(fn, deadline):
    while True:
        try:
            return fn()
        except Exception as e:
            if not is_contended(e) or time.time() >= deadline:
                raise
            sleep(0.05)
